### PR TITLE
add usage for expression filter

### PR DESCRIPTION
**Pull request by Liuxiaozhen12**
### What is changed, added, or deleted? (Required)
add doc for https://github.com/pingcap/dm/pull/1782

### Which DM version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [x] v2.0 (TiDB DM 2.0 versions)
- [ ] v1.0 (TiDB DM 1.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/docs-dm/pull/729
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: en/feature-expression-filter.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    expression-filters: ["even_c"]
+
+expression-filter:
+  even_c:
+    schema: "expr_filter"
+    table: "tbl"
+    insert-value-expr: "c % 2 = 0"
+```
+
+The above example configures `even_c` rule, and allows the data source whose ID is `mysql-replica-01` to refer this rule.The meaning of `even_c` is:
```
Comment (lance6716) [inline en/feature-expression-filter.md:42]:
```suggestion
The above example configures `even_c` rule, and allows the data source whose ID is `mysql-replica-01` to refer this rule. The meaning of `even_c` is:
```

File: en/feature-expression-filter.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
++------+------+
+2 rows in set (0.001 sec)
+```
+
+## Configuration parameters and rule descriptions
+
+- `schema`: The name of upstream database to be matched. Wildcard match or regular match is not supported.
+- `table`: The name of upstream table to be matched. Wildcard match or regular match is not supported.
+- `insert-value-expr`: To configure an expression which takes effect on the value of binlog event (WRITE_ROWS_EVENT) of INSERT type. It cannot appear in a same configuration item with `update-old-value-expr`, `update-new-value-expr`, and `delete-value-expr`.
+- `update-old-value-expr`：To configure an expression which takes effect on the old value of binlog event (UPDATE_ROWS_EVENT)  of UPDATE type. It cannot appear in a same configuration item with `insert-value-expr` and `delete-value-expr`.
```
Comment (lance6716) [inline en/feature-expression-filter.md:72]:
```suggestion
- `update-old-value-expr`：To configure an expression which takes effect on the old value of binlog event (UPDATE_ROWS_EVENT) of UPDATE type. It cannot appear in a same configuration item with `insert-value-expr` and `delete-value-expr`.
```

File: en/feature-expression-filter.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+>
+> - When you configure `update-old-value-expr` and `update-new-value-expr` at the same time, the row changes where updated old value meets the rule of `update-old-value-expr` **and** the updated new value meets the rule of `update-new-value-expr` will be filtered out.
+> - When you only configure one parameter, the statement you configure will decide whether to filter **the whole row changes**, which means the delete event of an old value and the insert event of a new value will be filtered out as a whole.
+
+SQL expressions can involve one or more columns. You can also use the SQL functions TiDB supports, such as `c % 2 = 0`, `a*a + b*b = c*c`, and `ts > NOW()`.
+
+The timezone of TIMESTAMP is UTC by default. You can use `c_timestamp = '2021-01-01 12:34:56.5678+08:00'` to show the specified timezone.
+
+You can define multiple filter rules under the configuration item `expression-filter`. By refering the rules you need in the configuration item of `expression-filters` in the upstream data source, the rules can take effect. When multiple rules take effect, matching **any** zh/task-configuration-file-full.mdof the rules cause a row change to be filtered.
```
Comment (lance6716) [inline en/feature-expression-filter.md:87]: not URLed

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: en/feature-expression-filter.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
```
Comment (Liuxiaozhen12) [inline en/feature-expression-filter.md:87]:
```suggestion
You can define multiple filter rules under the configuration item `expression-filter`. By refering the rules you need in the configuration item of `expression-filters` in the upstream data source, the rules can take effect. When multiple rules take effect, matching **any** of the rules cause a row change to be filtered.
```

File: en/feature-expression-filter.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+>
+> - When you configure `update-old-value-expr` and `update-new-value-expr` at the same time, the row changes where updated old value meets the rule of `update-old-value-expr` **and** the updated new value meets the rule of `update-new-value-expr` will be filtered out.
+> - When you only configure one parameter, the statement you configure will decide whether to filter **the whole row changes**, which means the delete event of an old value and the insert event of a new value will be filtered out as a whole.
+
+SQL expressions can involve one or more columns. You can also use the SQL functions TiDB supports, such as `c % 2 = 0`, `a*a + b*b = c*c`, and `ts > NOW()`.
+
+The timezone of TIMESTAMP is UTC by default. You can use `c_timestamp = '2021-01-01 12:34:56.5678+08:00'` to show the specified timezone.
+
+You can define multiple filter rules under the configuration item `expression-filter`. By refering the rules you need in the configuration item of `expression-filters` in the upstream data source, the rules can take effect. When multiple rules take effect, matching **any** of the rules cause a row change to be filtered.
```
Comment (Liuxiaozhen12) [inline en/feature-expression-filter.md:87]:
```suggestion
You can define multiple filter rules under the configuration item `expression-filter`. By refering the rules you need in the configuration item of `expression-filters` in the upstream data source, the rules can take effect. When multiple rules take effect, matching **any** of the rules causes a row change to be filtered.
```

**Review by lance6716 — APPROVED** [review]

File: en/feature-expression-filter.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+title: Filter Certain Row Changes Using SQL Expressions
+---
+
+# Filter Certain Row Changes Using SQL Expressions
+
+## Overview
+
+In the process of data migration, DM provides the [Binlog Event Filter](key-features.md#binlog-event-filter) feature to filter certain types of binlog events. For example, `DELETE` event might be filtered when data is migrated to the downstream for archiving or auditing, etc. However, Binlog Event Filter feature cannot judge more specifically whether the `DELETE` event of a certain row should be filtered.
+
+To solve the above issue, DM supports filtering certain row changes using SQL expressions. The binlog of ROW mode required by DM have the values of all columns in binlog event. You can configure SQL expressions according to these values. If the SQL expressions evaluate a row change as `TRUE`, DM will not migrate the row change downstream.
```
Comment (hfxsd) [inline en/feature-expression-filter.md:11]: 下面这句话的对应的中文看不懂，中英文可能都需要再优化一下。
The binlog of ROW mode required by DM have the values of all columns in binlog event.

File: en/feature-expression-filter.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
+---
+title: Filter Certain Row Changes Using SQL Expressions
+---
+
+# Filter Certain Row Changes Using SQL Expressions
+
+## Overview
+
+In the process of data migration, DM provides the [Binlog Event Filter](key-features.md#binlog-event-filter) feature to filter certain types of binlog events. For example, `DELETE` event might be filtered when data is migrated to the downstream for archiving or auditing, etc. However, Binlog Event Filter feature cannot judge more specifically whether the `DELETE` event of a certain row should be filtered.
```
Comment (hfxsd) [inline en/feature-expression-filter.md:9]:
```suggestion
In the process of data migration, DM provides the [Binlog Event Filter](key-features.md#binlog-event-filter) feature to filter certain types of binlog events. For example, for archiving or auditing purposes, `DELETE` event might be filtered when data is migrated to the downstream. However, Binlog Event Filter cannot judge with a greater granularity whether the `DELETE` event of a certain row should be filtered.
```

File: en/feature-expression-filter.md [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+To solve the above issue, DM supports filtering certain row changes using SQL expressions. The binlog of ROW mode required by DM have the values of all columns in binlog event. You can configure SQL expressions according to these values. If the SQL expressions evaluate a row change as `TRUE`, DM will not migrate the row change downstream.
+
+> **Note:**
+>
+> This feature only takes effect in the phase of incremental replication, not in the phase of full migration.
+
+## Configuration example
+
+Like [Binlog Event Filter](key-features.md#binlog-event-filter), you also need to configure the expression-filter feature in the configuration file of data migration task, as shown below. For complete configuration and its meaning, refer to [DM Advanced Task Configuration File](task-configuration-file-full.md#task-configuration-file-template-advanced)：
```
Comment (hfxsd) [inline en/feature-expression-filter.md:19]:
```suggestion
Similar to [Binlog Event Filter](key-features.md#binlog-event-filter), you also need to configure the expression-filter feature in the configuration file of the data migration task, as shown below. For complete configuration and its descriptions, refer to [DM Advanced Task Configuration File](task-configuration-file-full.md#task-configuration-file-template-advanced)：
```

File: en/feature-expression-filter.md [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
++------+------+
+|    1 |    1 |
+|    3 |    3 |
++------+------+
+2 rows in set (0.001 sec)
+```
+
+## Configuration parameters and rule descriptions
+
+- `schema`: The name of upstream database to be matched. Wildcard match or regular match is not supported.
```
Comment (hfxsd) [inline en/feature-expression-filter.md:69]:
```suggestion
- `schema`: The name of the upstream database to be matched. Wildcard match or regular match is not supported.
```

File: en/feature-expression-filter.md [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+- `table`: The name of upstream table to be matched. Wildcard match or regular match is not supported.
```
Comment (hfxsd) [inline en/feature-expression-filter.md:70]:
```suggestion
- `table`: The name of the upstream table to be matched. Wildcard match or regular match is not supported.
```

File: en/feature-expression-filter.md [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+- `insert-value-expr`: To configure an expression which takes effect on the value of binlog event (WRITE_ROWS_EVENT) of INSERT type. It cannot appear in a same configuration item with `update-old-value-expr`, `update-new-value-expr`, and `delete-value-expr`.
```
Comment (hfxsd) [inline en/feature-expression-filter.md:71]: 1 需要确认，不能同时和三个表达式使用，还是不能和三个表达式中的任意一个同时使用。
2 It cannot appear 那句话建议改成 
Do not use it with  `update-old-value-expr`, `update-new-value-expr`, and `delete-value-expr` in the same configuration item.
如果是不能与任意一个同时使用，要改成：
Do not use it with  `update-old-value-expr`, `update-new-value-expr`, or `delete-value-expr` in the same configuration item.
3 To configure 都改成 Specifies 
4 下面其他几个描述也都同步更新下。

File: en/feature-expression-filter.md [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+> **Note:**
+>
+> You can configure `update-old-value-expr` and `update-new-value-expr` at the same time.
+>
+> - When you configure `update-old-value-expr` and `update-new-value-expr` at the same time, the row changes where updated old value meets the rule of `update-old-value-expr` **and** the updated new value meets the rule of `update-new-value-expr` will be filtered out.
+> - When you only configure one parameter, the statement you configure will decide whether to filter **the whole row changes**, which means the delete event of an old value and the insert event of a new value will be filtered out as a whole.
+
+SQL expressions can involve one or more columns. You can also use the SQL functions TiDB supports, such as `c % 2 = 0`, `a*a + b*b = c*c`, and `ts > NOW()`.
+
+The timezone of TIMESTAMP is UTC by default. You can use `c_timestamp = '2021-01-01 12:34:56.5678+08:00'` to show the specified timezone.
```
Comment (hfxsd) [inline en/feature-expression-filter.md:85]: 中文是 显式指定时区，不是 显示指定时区。有误解。修改如下
```suggestion
The timezone of TIMESTAMP is UTC by default. You can use `c_timestamp = '2021-01-01 12:34:56.5678+08:00'` to specify the timezone explicitly.
```

File: en/feature-expression-filter.md [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+You can define multiple filter rules under the configuration item `expression-filter`. By refering the rules you need in the configuration item of `expression-filters` in the upstream data source, the rules can take effect. When multiple rules take effect, matching **any** of the rules causes a row change to be filtered.
+
+> **Note:**
+>
+> Setting too many expression filters for a table increases the computing overhead of DM， which might slow down data migration.
```
Comment (hfxsd) [inline en/feature-expression-filter.md:91]:
```suggestion
> Setting too many expression filters for a table increases the computing overhead of DM， which might impede data migration.
```

File: en/overview.md [hunk 14]
```diff
@@ -49,6 +49,10 @@ DM supports merging and migrating the original sharded instances and tables from
 
 In the MySQL ecosystem, tools such as gh-ost and pt-osc are widely used. DM provides support for these tools to avoid migrating unnecessary intermediate data. For details, see [Online DDL Tools](key-features.md#online-ddl-tools)
 
+### Filter certain row changes using SQL expressions
+
+In the phase of incremental replication, DM supports the configuration of SQL expressions to filter out certain row changes, which makes you more specifically control the data to be replicated. For more detailed information, refer to [Filter Certain Row Changes Using SQL Expressions](feature-expression-filter.md).
```
Comment (hfxsd) [inline en/overview.md:54]:
```suggestion
In the phase of incremental replication, DM supports the configuration of SQL expressions to filter out certain row changes, which lets you replicate the data with a greater granularity. For more information, refer to [Filter Certain Row Changes Using SQL Expressions](feature-expression-filter.md).
```

File: en/task-configuration-file-full.md [hunk 15]
```diff
@@ -78,6 +78,13 @@ filters:
...
 
+expression-filter:                   # Defines the filter rules for row changes when migrating data. Supports defining multiple rules.
+  # Filter the value of inserted `c` in `expr_filter`.`tbl` when it is even.
+  even_c:                            # The name of the filter rule.
+    schema: "expr_filter"            # The name of upstream database to be matched. Wildcard match or regular match is not supported.
+    table: "tbl"                     # The name of upstream table to be matched. Wildcard match or regular match is not supported.
+    insert-value-expr: "c % 2 = 0"
+
 # The filter rule set of tables to be migrated from the upstream database instance. You can set multiple rules at the same time.
 block-allow-list:                    # Use black-white-list if the DM version is earlier than v2.0.0-beta.2.
```
Comment (hfxsd) [inline en/task-configuration-file-full.md:89]:
```suggestion
block-allow-list:                    # Use black-white-list if the DM version is earlier than or equal to v2.0.0-beta.2.
```

File: en/task-configuration-file-full.md [hunk 16]
```diff
@@ -125,8 +132,9 @@ mysql-instances:
       binlog-gtid: "03fc0263-28c7-11e7-a653-6c0b84d59f30:1-7041423,05474d3c-28c7-11e7-8352-203db246dd3d:1-170"  # You need to set this argument if you specify `enable-gtid: true` for the source of the incremental task.
 
     route-rules: ["route-rule-1", "route-rule-2"]   # The name of the mapping rule between the table matching the upstream database instance and the downstream database.
-    filter-rules: ["filter-rule-1", "filter-rule-2"]                 # The name of the binlog event filtering rule of the table matching the upstream database instance.
+    filter-rules: ["filter-rule-1", "filter-rule-2"] # The name of the binlog event filtering rule of the table matching the upstream database instance.
     block-allow-list:  "bw-rule-1"                  # The name of the block and allow lists filtering rule of the table matching the upstream database instance. Use black-white-list if the DM's version <= v2.0.0-beta.2.
```
Comment (hfxsd) [inline en/task-configuration-file-full.md:136]:
```suggestion
    block-allow-list:  "bw-rule-1"                  # The name of the block and allow lists filtering rule of the table matching the upstream database instance. Use black-white-list if the DM version is earlier than or equal to v2.0.0-beta.2.
```

File: en/task-configuration-file-full.md (hunk 15 above)
Comment (Liuxiaozhen12) [reply en/task-configuration-file-full.md:89]: 之前的 PR 已修改过，等之前的 PR 合并后才能看到修改后的内容，这里就不用重复改啦~

File: en/task-configuration-file-full.md (hunk 16 above)
Comment (Liuxiaozhen12) [reply en/task-configuration-file-full.md:136]: ditto

File: en/feature-expression-filter.md (hunk 6 above)
Comment (Liuxiaozhen12) [reply en/feature-expression-filter.md]: addressed in https://github.com/pingcap/docs-dm/pull/753/commits/260ec99d6dffe909ef9bf373df797d4fa4c1a05a.

**Comment by hfxsd** [comment]
LGTM

**Review by TomShawn — APPROVED** [review]

File: en/feature-expression-filter.md [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+title: Filter Certain Row Changes Using SQL Expressions
+---
+
+# Filter Certain Row Changes Using SQL Expressions
+
+## Overview
+
+In the process of data migration, DM provides the [Binlog Event Filter](key-features.md#binlog-event-filter) feature to filter certain types of binlog events. For example, for archiving or auditing purposes, `DELETE` event might be filtered when data is migrated to the downstream. However, Binlog Event Filter cannot judge with a greater granularity whether the `DELETE` event of a certain row should be filtered.
+
+To solve the above issue, DM supports filtering certain row changes using SQL expressions. The binlog in the `ROW` format supported by DM has the values of all columns in binlog event. You can configure SQL expressions according to these values. If the SQL expressions evaluate a row change as `TRUE`, DM will not migrate the row change downstream.
```
Comment (Liuxiaozhen12) [inline en/feature-expression-filter.md:11]:
```suggestion
To solve the above issue, DM supports filtering certain row changes using SQL expressions. The binlog in the `ROW` format supported by DM has the values of all columns in binlog events. You can configure SQL expressions according to these values. If the SQL expressions evaluate a row change as `TRUE`, DM will not migrate the row change downstream.
```

**Pull request merged**